### DISCUSSION TITLE
Doublehop VPN

### DISCUSSION THREAD

**Discussion by Ilya87** [strongswan/strongswan#985] [Q&A]
Hello! I have two VPS (vps1 and vps2 with). Both have Stronswan based ikev2 ipsec VPN server that allow clients to connect and use internet from the ip-address of vps.
**/etc/ipsec.conf**
```
config setup
    charondebug="dmn 0, mgr 0, ike 0, chd 0, job 0, cfg 0, knl 0, net 0, asn 0, enc 0, lib 0, esp 0, tls 0, tnc 0, imc 0, imv 0, pts 0"
    uniqueids=never
conn VPS1-ikev2-vpn
    auto=route
    compress=no
    type=tunnel
    keyexchange=ikev2
    fragmentation=yes
    forceencaps=yes
    dpdaction=clear
    dpddelay=5s
    rekey=no
    left=%any
    leftid=VPS1_STATIC_IP
    leftcert=server-cert.pem
    leftsendcert=always
    leftsubnet=0.0.0.0/0
    right=%any
    rightid=%any
    rightauth=eap-mschapv2
    rightsourceip=10.10.10.0/24
    rightdns=1.1.1.1,8.8.8.8
    rightsendcert=never
    eap_identity=%identity
    ike=chacha20poly1305-sha256-curve25519-prfsha256,aes256gcm16-sha256-prfsha256-ecp256!
    esp=chacha20poly1305-sha256,aes256gcm16-ecp256,aes256-sha256!
    keyingtries=%forever
    lifetime=24h
    closeaction=clear
```
I have found no solution how to create doublehop VPN so that vps2 connects to vps1 using ikev2 ipsec and all the traffic from clients, connected to vps2 forwards to vps1 and vps1 is a gateway to the Internet.
Help, please!

**Comment by tobiasbrunner** [comment]
Use distinct subnets for `rightsourceip` and then create a tunnel between the subnet used on vps2 and `0.0.0.0/0` between the two hosts (the NAT on vps1 has to handle both subnets). Alternatively, NAT traffic from clients to an IP on vps2 and create a tunnel to vps1 for traffic from that IP to `0.0.0.0/0` (again the NAT on vps1 has to handle that traffic). Other than that, read [this](https://docs.strongswan.org/strongswan-docs/5.9/howtos/forwarding.html) for general information on forwarding traffic.

**Comment by Ilya87** [comment]
VPS1 added block to /etc/ipsec.conf
```conn vps1-to-vps2
  authby=secret
  left=%defaultroute
  leftid=VPS1_IP
  leftsubnet=10.10.10.0/24
  right=VPS2_IP
  rightsubnet=10.10.9.0/24
  ike=chacha20poly1305-sha256-curve25519-prfsha256,aes256gcm16-sha256-prfsha256-ecp256!
  esp=chacha20poly1305-sha256,aes256gcm16-ecp256,aes256-sha256!
  keyingtries=0
  ikelifetime=1h
  lifetime=8h
  dpddelay=30
  dpdtimeout=120
  dpdaction=restart
  auto=start
```  

VPS2 added block to /etc/ipsec.conf
```
conn vps2-to-vps1
  authby=secret
  left=%defaultroute
  leftid=VPS2_IP
  leftsubnet=10.10.9.0/24
  right=VPS1_IP
  rightsubnet=10.10.10.0/24
  ike=chacha20poly1305-sha256-curve25519-prfsha256,aes256gcm16-sha256-prfsha256-ecp256!
  esp=chacha20poly1305-sha256,aes256gcm16-ecp256,aes256-sha256!
  keyingtries=0
  ikelifetime=1h
  lifetime=8h
  dpddelay=30
  dpdtimeout=120
  dpdaction=restart
  auto=start
```

**/etc/ufw/before.rules** (vps1, vps 2 has the same rules, the only difference is subnet 10.10.9.0/24)
```
*nat
-A POSTROUTING -s 10.10.10.0/24 -o ens3 -m policy --pol ipsec --dir out -j ACCEPT
-A POSTROUTING -s 10.10.10.0/24 -o ens3 -j MASQUERADE
COMMIT

*mangle
-A FORWARD --match policy --pol ipsec --dir in -s 10.10.10.0/24 -o ens3 -p tcp -m tcp --tcp-flags SYN,RST SYN -m tcpmss --mss 1361:1536 -j TCPMSS --set-mss 1360
COMMIT
# Don't delete these required lines, otherwise there will be errors
*filter
:ufw-before-input - [0:0]
:ufw-before-output - [0:0]
:ufw-before-forward - [0:0]
:ufw-not-local - [0:0]
# End required lines
-A ufw-before-forward --match policy --pol ipsec --dir in --proto esp -s 10.10.10.0/24 -j ACCEPT
-A ufw-before-forward --match policy --pol ipsec --dir out --proto esp -d 10.10.10.0/24 -j ACCEPT
```
Is my config correct? And what rule for NAT for vps1 should be added?

**Reply by tobiasbrunner** [reply]
> Is my config correct?

No. Your config only allows traffic between the subnets `10.10.9.0/24` and `10.10.10.0/24` between the two gateways, it does not allow one gateway to forward traffic addressed to any host on the internet that it receives from its clients.

You want `rightsubnet` on vps2 to be `0.0.0.0/0` (likewise for `leftsubnet` on vps1). Then vps2 can forward traffic from its clients to vps1, where you want to treat it like traffic directly received from clients (just the source subnet is different).

> And what rule for NAT for vps1 should be added?

Basically the same you already have for `10.10.10.0/24` but also for `10.10.9.0/24`. And you don't need NAT rules on vps2 as that traffic should just get forwarded to vps1.

**Comment by Ilya87** [comment]
The connection is establshed, but still no internet on clients of vps2. Yes, I know, I'm not good at iptables rules.
**vps2 iptables rules**

```
*mangle
-A FORWARD --match policy --pol ipsec --dir in -s 10.10.9.0/24 -o ens3 -p tcp -m tcp --tcp-flags SYN,RST SYN -m tcpmss --mss 1361:1536 -j TCPMSS --set-mss 1360
COMMIT
# Don't delete these required lines, otherwise there will be errors
*filter
:ufw-before-input - [0:0]
:ufw-before-output - [0:0]
:ufw-before-forward - [0:0]
:ufw-not-local - [0:0]
# End required lines
-A ufw-before-forward --match policy --pol ipsec --dir in --proto esp -s 10.10.9.0/24 -j ACCEPT
-A ufw-before-forward --match policy --pol ipsec --dir out --proto esp -d 10.10.9.0/24 -j ACCEPT
```

**vps1 iptables rules**

```
*nat
-A POSTROUTING -s 10.10.10.0/24 -o ens3 -m policy --pol ipsec --dir out -j ACCEPT
-A POSTROUTING -s 10.10.10.0/24 -o ens3 -j MASQUERADE
-A POSTROUTING -s 10.10.9.0/24 -o ens3 -m policy --pol ipsec --dir out -j ACCEPT
-A POSTROUTING -s 10.10.9.0/24 -o ens3 -j MASQUERADE
COMMIT

*mangle
-A FORWARD --match policy --pol ipsec --dir in -s 10.10.10.0/24 -o ens3 -p tcp -m tcp --tcp-flags SYN,RST SYN -m tcpmss --mss 1361:1536 -j TCPMSS --set-mss 1360
-A FORWARD --match policy --pol ipsec --dir in -s 10.10.9.0/24 -o ens3 -p tcp -m tcp --tcp-flags SYN,RST SYN -m tcpmss --mss 1361:1536 -j TCPMSS --set-mss 1360
COMMIT
# Don't delete these required lines, otherwise there will be errors
*filter
:ufw-before-input - [0:0]
:ufw-before-output - [0:0]
:ufw-before-forward - [0:0]
:ufw-not-local - [0:0]
# End required lines
-A ufw-before-forward --match policy --pol ipsec --dir in --proto esp -s 10.10.10.0/24 -j ACCEPT
-A ufw-before-forward --match policy --pol ipsec --dir out --proto esp -d 10.10.10.0/24 -j ACCEPT
-A ufw-before-forward --match policy --pol ipsec --dir in --proto esp -s 10.10.9.0/24 -j ACCEPT
-A ufw-before-forward --match policy --pol ipsec --dir out --proto esp -d 10.10.9.0/24 -j ACCEPT
```

**Comment by Ilya87** [comment]
-A POSTROUTING -s 10.0.10.0/24 -o ens3 -j SNAT --to-source vps1_ip
And it is done! 
P.S.
NOOO!

**Reply by tobiasbrunner** [reply]
Hm, what has `10.0.10.0/24` to do with your config? Or is that a typo? And why wouldn't it work with `MASQUERADE`?

**Comment by Ilya87** [comment]
-A POSTROUTING -s 10.0.10.0/24 -o ens3 -j SNAT --to-source vps1_ip is not needed
I have found out (maybe) the reason - ufw. All the iptables rules are in /etc/ufw/before.rules. So I need `ufw disable && ufw enable && ipsec restart`. No it still doesn't work
[before.rules.txt](https://github.com/strongswan/strongswan/files/8417375/before.rules.txt)
[vps1.txt](https://github.com/strongswan/strongswan/files/8417817/vps1.txt)
[vps2.txt](https://github.com/strongswan/strongswan/files/8417818/vps2.txt)
